include/hydra/places/traversability_estimator.h

```diff
@@ -0,0 +1,132 @@
+  using ConstPtr = std::shared_ptr<const TraversabilityEstimator>;
+
+  TraversabilityEstimator() = default;
+  virtual ~TraversabilityEstimator() = default;
+
+  /**
+   * @brief Update the traversability based on the active window output and mesh delta.
+   * @param msg The active window output containing the latest sensor data and
+   * volumetric map updates.
+   * @param mesh_delta The mesh delta representing changes in the mesh.
+   * @param graph The dynamic scene graph to update with traversability information.
+   */
+  virtual void updateTraversability(const ActiveWindowOutput& msg,
+                                    const kimera_pgmo::MeshDelta& mesh_delta,
+                                    const spark_dsg::DynamicSceneGraph& graph) = 0;
+
+  const TraversabilityLayer& getTraversabilityLayer() const {
+    return *traversability_layer_;
+  }
+
+ protected:
+  std::unique_ptr<TraversabilityLayer> traversability_layer_;
+};
+
+/**
+ * @brief Simple traversability estimator which checks a specified volume in the TSDF
+ * map for obstacles.
+ * @note Assumptions: Only works for flat ground, and does not consider cliffs as
+ * obstacles.
+ */
+class HeightTraversabilityEstimator : public TraversabilityEstimator {
+ public:
+  struct Config {
+    //! @brief The height above the robot body to consider for traversability in meters.
+    float height_above = 0.5f;
+
+    //! @brief The height below the robot body to consider for traversability in meters.
+    float height_below = 0.5f;
+
+    //! @brief Minimum confidence for a voxel to be considered observed.
+    float min_confidence = 1.0f;
+
+    //! @brief Minimum traversability for a voxel to be considered traversable.
+    float min_traversability = 1.0f;
+
+    //! @brief If true, mark voxels as intraversable if they do not meet the
+    //! min_traversability threshold, even if the confidence is below min_confidence. If
+    //! false, mark these voxels as unknown instead.
+    bool pessimistic = true;
+  };
+
+  HeightTraversabilityEstimator(const Config& config);
+  ~HeightTraversabilityEstimator() override = default;
+
+  void updateTraversability(const ActiveWindowOutput& msg,
+                            const kimera_pgmo::MeshDelta& /* mesh_delta */,
+                            const spark_dsg::DynamicSceneGraph& /* graph */) override;
+
+  const Config config;
+
+ protected:
+  TsdfLayer::Ptr tsdf_layer_;
+
+  // Processing steps.
+  void updateTsdf(const ActiveWindowOutput& msg);
+  void computeTraversability(const ActiveWindowOutput& msg);
+  void classifyTraversabilityVoxel(TraversabilityVoxel& voxel) const;
+
+  // Helper functions.
+  BlockIndexSet get2DBlockIndices(const BlockIndices& blocks) const;
+
+ private:
+  inline static const auto registration_ =
+      config::RegistrationWithConfig<TraversabilityEstimator,
+                                     HeightTraversabilityEstimator,
+                                     Config>("HeightTraversabilityEstimator");
```

Comment:
(minor) better to define registrations on the implementation side, the `config_utilities/factory.h` header is expensive to pull in compilation-wise and you get a bunch of multiply-defined registration errors for the python bindings. I think the actual condition for the errors is when the core Hydra library and the bindings both include the header with the registration, but that's hard to predict so I've been moving everything to the source side when I see it.